### PR TITLE
IssueProcessor integration tests.

### LINKED ISSUE #375: Create integration test for @Issue extensions
While trying to write an integration test for the `IssueExtension` it seems to turn out that the `TestKitEngine` does not pick up the globally registered `TestExecutionListener` and delegates published `ReportEntries` to it.

Maybe useful links:
https://github.com/junit-team/junit5/pull/2006/files

### PR DESCRIPTION
**Pull request by Michael1993**
Closes #375 

**PR checklist**

The following checklist shall help the PR's author, the reviewers and maintainers to ensure the quality of this project.
It is based on our contributors guidelines, especially the ["writing code" section](https://github.com/junit-pioneer/junit-pioneer/blob/master/CONTRIBUTING.md#writing-code).
It shall help to check for completion of the listed points.
If a point does not apply to the given PR's changes, the corresponding entry can be simply marked as done. 

Documentation (general)
* [ ] There is documentation (Javadoc and site documentation; added or updated)
* [ ] There is implementation information to describe _why_ a non-obvious source code / solution got implemented
* [ ] Site documentation has its own `.adoc` file in the `docs` folder, e.g. `docs/report-entries.adoc`
* [ ] Only one sentence per line (especially in `.adoc` files)
* [ ] Javadoc uses formal style, while sites documentation may use informal style

Documentation (new extension)
* [ ] The `docs/docs-nav.yml` navigation has an entry for the new extension
* [ ] The `package-info.java` contains information about the new extension

Code
* [ ] Code adheres to code style, naming conventions etc.
* [ ] Successful tests cover all changes
* [ ] There are checks which validate correct / false usage / configuration of a functionality and there are tests to verify those checks
* [ ] Tests use [AssertJ](https://joel-costigliola.github.io/assertj/) or our own [PioneerAssert](https://github.com/junit-pioneer/junit-pioneer/blob/master/CONTRIBUTING.md#assertions) (which are based on AssertJ)

Contributing
* [ ] A prepared commit message exists
* [ ] The list of contributions inside `README.md` mentions the new contribution (real name optional) 

---

I hereby agree to the terms of the [JUnit Pioneer Contributor License Agreement](https://github.com/junit-pioneer/junit-pioneer/blob/master/CONTRIBUTING.md#junit-pioneer-contributor-license-agreement).

### CODE REVIEW

**Comment by Michael1993** [comment]
@Bukama Feel free to check out and use this branch as baseline for the integration tests - or I could just add a bunch of assertions (?).

File: src/main/java/org/junitpioneer/jupiter/issue/IssueExtensionExecutionListener.java [hunk 1]
```diff
@@ -37,7 +36,7 @@
  */
 public class IssueExtensionExecutionListener implements TestExecutionListener {
 
-	public static final String REPORT_ENTRY_KEY = "Issue";
+	public static final String REPORT_ENTRY_KEY = IssueExtensionExecutionListener.class.toString();
```
Comment (Bukama) [inline src/main/java/org/junitpioneer/jupiter/issue/IssueExtensionExecutionListener.java:39]: Hmm. We define the report key in all extensions etc. with a string and not by the name of a class. Don't know if I like this change

File: src/main/java/org/junitpioneer/jupiter/issue/IssueProcessorFactory.java [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+ * http://www.eclipse.org/legal/epl-v20.html
+ */
+
+package org.junitpioneer.jupiter.issue;
+
+import java.util.ServiceLoader;
+
+import org.junitpioneer.jupiter.IssueProcessor;
+
+// Ensures we get the same instance of StoringIssueProcessor in the integration tests
```
Comment (Bukama) [inline src/main/java/org/junitpioneer/jupiter/issue/IssueProcessorFactory.java:17]: No correct Javadoc ;)

File: src/test/java/org/junitpioneer/jupiter/issue/IssueExtensionIntegrationTests.java [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+		@Test
+		@Disabled("skipped")
+		@Issue("When you awaken in the morning's hush, I am the swift uplifting rush")
+		void skippedTest() {
+		}
+
+		@Test
+		@Issue("Of quiet birds in circled flight. I am the soft stars that shine at night.")
+		void publishingTest(TestReporter reporter) {
+			reporter.publishEntry("Issue", "reporting test");
```
Comment (Bukama) [inline src/test/java/org/junitpioneer/jupiter/issue/IssueExtensionIntegrationTests.java:81]: Do you want to use the new key here?

File: src/test/java/org/junitpioneer/jupiter/issue/IssueExtensionIntegrationTests.java [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+
+	private final StoringIssueProcessor issueProcessor = (StoringIssueProcessor) IssueProcessorFactory
+			.getIssueProcessors()
+			.iterator()
+			.next();
+
+	@Test
+	void testIssueCases() {
+		LauncherFactory
+				.create()
```
Comment (Bukama) [inline src/test/java/org/junitpioneer/jupiter/issue/IssueExtensionIntegrationTests.java:42]: What does this?

File: src/test/java/org/junitpioneer/jupiter/issue/IssueExtensionIntegrationTests.java [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+import org.junitpioneer.jupiter.Issue;
+import org.junitpioneer.jupiter.IssueTestSuite;
+import org.opentest4j.TestAbortedException;
+
+/**
+ * Mary Elizabeth Fyre: Do Not Stand at My Grave and Weep is in the public domain.
+ */
+public class IssueExtensionIntegrationTests {
+
+	private final StoringIssueProcessor issueProcessor = (StoringIssueProcessor) IssueProcessorFactory
```
Comment (Bukama) [inline src/test/java/org/junitpioneer/jupiter/issue/IssueExtensionIntegrationTests.java:34]: So you try to convert every IssueProcessoir which is found into a StoringIssueProcessor?

**Review by Bukama — COMMENTED** [review]
To be honest I'm a bit lost about what you all do.

File: src/test/java/org/junitpioneer/jupiter/issue/IssueExtensionIntegrationTests.java (hunk 4 above)
Comment (Michael1993) [reply src/test/java/org/junitpioneer/jupiter/issue/IssueExtensionIntegrationTests.java:42]: Creates a [`Launcher`](https://junit.org/junit5/docs/5.0.3/api/org/junit/platform/launcher/Launcher.html), which gets provided with a [`LauncherDiscoveryRequest`](https://junit.org/junit5/docs/5.0.3/api/org/junit/platform/launcher/LauncherDiscoveryRequest.html) and executes it.

Essentially indirectly executes the `IssueIntegrationCases` class. By default, the `LauncherFactory` will use `ServiceLoader` to load in `TestExecutionListener`s (so `IssueExtensionExecutionListener` is automatically loaded in during the execution).

File: src/test/java/org/junitpioneer/jupiter/issue/IssueExtensionIntegrationTests.java (hunk 5 above)
Comment (Michael1993) [reply src/test/java/org/junitpioneer/jupiter/issue/IssueExtensionIntegrationTests.java:34]: It's our own tests, so I made the bold assumption that we will only use the one `IssueProcessor`. 😛

Comment (Michael1993) [reply src/test/java/org/junitpioneer/jupiter/issue/IssueExtensionIntegrationTests.java:34]: I'm fairly certain this won't change in the future, but if it does we can update the test?

File: src/test/java/org/junitpioneer/jupiter/issue/IssueExtensionIntegrationTests.java (hunk 3 above)
Comment (Michael1993) [reply src/test/java/org/junitpioneer/jupiter/issue/IssueExtensionIntegrationTests.java:81]: No, I specifically used the "Issue" string to see what happens. It's not good - the issue id gets overwritten by whatever gets published with "Issue" key. 😞

File: src/main/java/org/junitpioneer/jupiter/issue/IssueProcessorFactory.java (hunk 2 above)
Comment (Michael1993) [reply src/main/java/org/junitpioneer/jupiter/issue/IssueProcessorFactory.java:17]: Does this class need javadoc? It's entirely internal to Pioneer.

File: src/main/java/org/junitpioneer/jupiter/issue/IssueExtensionExecutionListener.java (hunk 1 above)
Comment (Michael1993) [reply src/main/java/org/junitpioneer/jupiter/issue/IssueExtensionExecutionListener.java:39]: I don't _really_ like it either, but I think "Issue" is just a tad bit too generic. 🤷 

Perhaps this?
```
public static final String REPORT_ENTRY_KEY = "IssueExtensionExecutionListener";
```

Comment (Bukama) [reply src/main/java/org/junitpioneer/jupiter/issue/IssueExtensionExecutionListener.java:39]: I like this better, so we can refactor (if needed) the class without breaking "the API".

File: src/main/java/org/junitpioneer/jupiter/issue/IssueProcessorFactory.java (hunk 2 above)
Comment (Bukama) [reply src/main/java/org/junitpioneer/jupiter/issue/IssueProcessorFactory.java:17]: Well, you use the class in another place, so I think it's useful to add JavaDoc (aside from the fact I always add JavaDoc)

File: src/test/java/org/junitpioneer/jupiter/issue/IssueExtensionIntegrationTests.java (hunk 3 above)
Comment (Bukama) [reply src/test/java/org/junitpioneer/jupiter/issue/IssueExtensionIntegrationTests.java:81]: Allright.

File: src/main/java/org/junitpioneer/jupiter/issue/IssueProcessorFactory.java (hunk 2 above)
Comment (Michael1993) [reply src/main/java/org/junitpioneer/jupiter/issue/IssueProcessorFactory.java:17]: Our extensions don't have Javadoc either...? 🤷 I think only the public-facing API of Pioneer has proper javadoc.

Ahem. Sure, I can write a quick javadoc. No problemo.

**Review by Bukama — APPROVED** [review]
LGTM :)

**Comment by Michael1993** [comment]
I think I will write a few more assertions on the results and rename `IssueProcessorFactory` to `IssueProcessorRegistry`.

**Review by Bukama — APPROVED** [review]
I like the `IssueProcessorRegistry` - nice encapsulation :)

File: src/main/java/org/junitpioneer/jupiter/issue/IssueExtensionExecutionListener.java [hunk 6]
```diff
@@ -37,7 +36,7 @@
  */
 public class IssueExtensionExecutionListener implements TestExecutionListener {
 
-	public static final String REPORT_ENTRY_KEY = "Issue";
+	public static final String REPORT_ENTRY_KEY = "IssueExtensionExecutionListener";
```
Comment (nipafx) [inline src/main/java/org/junitpioneer/jupiter/issue/IssueExtensionExecutionListener.java:39]: This key is used by `IssueExtension` to publish report entries. Why would it use the key of the "target" of the message. If "Issue" is too general, let's stick to "Issue extension" (or "IssueExtension"`).

File: src/main/java/org/junitpioneer/jupiter/issue/IssueExtensionExecutionListener.java [hunk 7]
```diff
@@ -88,7 +87,7 @@ public void testPlanExecutionFinished(TestPlan testPlan) {
 			return;
 
 		List<IssueTestSuite> issueTestSuites = createIssueTestSuites();
-		for (IssueProcessor issueProcessor : ServiceLoader.load(IssueProcessor.class)) {
+		for (IssueProcessor issueProcessor : IssueProcessorRegistry.getIssueProcessors()) {
```
Comment (nipafx) [inline src/main/java/org/junitpioneer/jupiter/issue/IssueExtensionExecutionListener.java:90]: I don't like how these purely test-related problem bleeds into the production code.

**Review by nipafx — COMMENTED** [review]
Two things - will try to fix them myself.

**Pull request merged**